Week 8 Quiz App

MBHunter88 opened the pull request:
Week 8 Project (1st Batch of Commits)

-File Structure and Installation: Organized initial file structure and installed necessary packages (e.g., Express, React, Jest) (commits 1 & 2).
-Server Initialization: Initialized the server and successfully fetched data from the external Trivia API using Postman for testing (commit 3).
-Server Testing Setup: Set up server testing environment with Jest and Supertest. Faced challenges configuring Supertest to work with ES modules due to its CommonJS syntax. Resolved this issue with adjustments to import statements and configurations (commit 4).
-Proxy Configuration and Frontend Initialization: Configured the server proxy in vite.config.js file to handle API requests from the frontend. Initialized the Question component to handle data fetching (commit 5).
-Rendering Question and Answers: Rendered trivia question and answer selections  within the Question component, incorporating state management.  (commit 6).

Challenges Faced:

Difficulty configuring Supertest to support ES modules. Resolved this utilizing chatGpt, which involved modifying imports to handle CommonJS modules in an ES environment. I also would like to review the syntax for a deeper understanding. 

Rendering and Shuffling Answers: Faced challenges rendering data since each API request returns an array of 10 questions with separate object keys for incorrect and correct answers. To render the answers randomly, I combined all answers into a single array, shuffled the array, and then mapped each answer to a button for display.

Next Steps:

Implement scoring functionality and user feedback in the frontend.
Set up additional tests for frontend components using Jest and React Testing Library.

Commit:
first commit-server setup

Commit:
frontend set up

Commit:
initialize server and test api functionality

Commit:
add testing for api using jest

Commit:
connect back and front as well as initialize question component

Commit:
render answer selections as buttons in question component

Commit:
updated Question component to fetch 10 question at a time

Commit:
add game logic to change score if correct answer is selected by user

Commit:
restructured components for use with useContext hook

Commit:
add end game logic and improved Question component

Commit:
add commentation and changed variable names for clarity

Shikha2410 commented on techtonica-projects/quiz-app/client/vite.config.js:
```diff
@@ -0,0 +1,12 @@
+import { defineConfig } from 'vite'
+import react from '@vitejs/plugin-react'
+
+// https://vitejs.dev/config/
```
Can this be removed?

Shikha2410 commented on techtonica-projects/quiz-app/client/vite.config.js:
```diff
@@ -0,0 +1,12 @@
+  plugins: [react()],
+  server: {
+    proxy: {
+      '/api/trivia': 'http://localhost:3000', // Replace with your Express server's address
```
Let's move comment to its own line on top of `'/api/trivia': 'http://localhost:3000',` for better readability

Shikha2410 commented on techtonica-projects/quiz-app/server/server.js:
```diff
@@ -0,0 +1,41 @@
+import dotenv from "dotenv"
+
+
+//use .env for variables
```
An example of good & helpful documentation. Keep it up 👍

Shikha2410 commented on techtonica-projects/quiz-app/server/server.js:
```diff
@@ -0,0 +1,41 @@
+    console.log("Hello World!")
+})
+
+//create route for trivia api *note to later update request with a session token to avoid duplicate questions*
```
*note..* can be written on new line as follows:
`TODO: Update request with a session token to avoid duplicate questions`

That way you can run a single search with "TODO:" on your whole project in the end and make sure you've not missed any notes.

Shikha2410 commented on techtonica-projects/quiz-app/server/server.js:
```diff
@@ -0,0 +1,41 @@
+        const response = await fetch(apiUrl);
+        const data = await response.json();
+        res.json(data);
+    } catch { //include error catching
```
its called *error handling

Shikha2410 commented on techtonica-projects/quiz-app/server/server.js:
```diff
@@ -0,0 +1,41 @@
+//const urlencodedParser = bodyParser.urlencoded({ extended: false })
+
+
+//create route for test "homepage" or "root"
```
Do you need this route committed into project repo?

MBHunter88 replied:
No, now that I am implementing testing this route is no longer necessary and can be removed.

Shikha2410 commented on techtonica-projects/quiz-app/server/__test__/api.test.js:
```diff
@@ -0,0 +1,10 @@
+import request from 'supertest';
+import app from '../server.js'; 
+
+describe('GET /api/trivia', () => {
```
This seems incomplete. 
1. How are you planning to use the "request" you've imported?
2. Unit tests should generate test results as output when you run them in your local environment. Do include that result as is in your PR description.

MBHunter88 replied:
Thank you for the feedback! The syntax I used was based on my research to troubleshoot "supertest" and understand how to utilize Jest for testing API routes. I understand now that I should provide test results directly in the PR description. Moving forward, I hope to implement mocking to avoid making actual API requests during testing.

Shikha2410 commented on techtonica-projects/quiz-app/server/server.js:
```diff
@@ -38,4 +38,6 @@ app.get('/api/trivia', async (req, res) => {
\ No newline at end of file
+app.listen(PORT, () => console.log(`Server is runnning on port http://localhost:${PORT}`))
+
+export default app;
```
What were the other alternatives for fixing the issue you faced while configuring Supertest?
This export hack appears a bit of an overkill to me, maybe there's a simpler fix available?

MBHunter88 replied:
I added this line because I encountered an error that this syntax helped resolve, allowing the test to run successfully. I am continuing to troubleshoot and optimize the testing for the API.

```
    SyntaxError: The requested module '../server.js' does not provide an export named 'default'

Commit:
set up testing folder, add css styling, updates from mentor notes

Commit:
fixed special character display

Commit:
set successfully setup testing enviornments with test files in both server and client directories

Commit:
set successfully setup testing enviornments with test files in both server and client directories

Commit:
adjusted styling for accessibility and added media queries

Commit:
enhanced styling and UI

Commit:
added further testing for client directory

Commit:
final touches and README.md file

MBHunter88 commented:
Summary of Changes:

Testing Setup:

Created a dedicated __test__ folder.
Successfully configured testing environments with test files for both the server and client directories.
Added further unit and integration testing in the client directory to ensure seamless interactions between components.

Styling Enhancements:

Added enhanced CSS styling for a polished and modern UI.
Adjusted styling/comments/syntax based on mentor feedback to improve the overall user experience.
Ensured that special characters (like &, ") display correctly in trivia questions and answers.

Accessibility Improvements:

Made styling adjustments to meet accessibility standards, particularly with text color contrast.
Implemented media queries for responsive design across mobile, tablet, and desktop devices.

Final Touches:

Added a README.md file with installation, usage, and testing instructions.

Challenges Faced:

Setting up testing environments proved difficult due to a simple syntax error, but once resolved, I could complete simple tests throughout the project. 

Initially, I struggled with passing data and props between components. Upon reviewing the logic, I realized the app required additional components for better structure. Researching the useContext hook greatly simplified data passing, making the code more manageable and improving overall app organization.

Next Steps: Although I won't have time to implement any major changes I am looking forward to your feedback.

Commit:
added form to meet requirements, updated demo, ran tests to ensure functionality

Shikha2410 reviewed (approved):
Further discussed in 1:1. Good to merge.
